app/jobs/export_pool_xp_to_traction_job.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+# Combine the samples from a Pool XP tube into a compound sample, generate a bioscan-pool-xp-tube-to-traction message
+# and submit it to RabbitMQ so that it can forwarded to Traction by the message processor.
+ExportPoolXpToTractionJob =
+  Struct.new(:barcode) do
+    include CompoundSampleHelper
+
+    def perform
+      message_data = get_message_data(barcode)
+
+      subject_obj = configatron.amqp.schemas.subjects[:export_pool_xp_to_traction]
+      subject = subject_obj[:subject]
+      version = subject_obj[:version]
+
+      schema = get_message_schema(subject, version)
+      encoded_message = avro_encode_message(message_data, schema)
+      send_message(encoded_message, subject, version)
+    rescue StandardError => e
+      Rails.logger.error("Error exporting Pool XP tube to Traction: <#{e.message}>")
+      raise
+    end
+
+    def fetch(uri_str, limit = 10)
+      raise IOError, 'Too many HTTP redirects' if limit.zero?
+
+      response = Net::HTTP.get_response(URI.parse(uri_str))
+
+      case response
+        when Net::HTTPSuccess     then response
+        when Net::HTTPRedirection then fetch(response['location'], limit - 1)
+      else
+        response.error!
+      end
+    end
+
+    def get_message_data(barcode)
+      tube = Tube.with_barcode(barcode).first
+      project = tube.projects.first
+      study = tube.studies.first
+      sample = find_or_create_compound_sample(study, tube.samples)
+
+      {
+        messageUuid: UUIDTools::UUID.timestamp_create.to_s,
+        messageCreateDateUtc: Time.now.utc,
+        tubeBarcode: tube.human_barcode,
+        library: {
+          volume: 100.0,
+          concentration: 0.0,
+          boxBarcode: "Unspecified"
+        },
+        request: {
+          costCode: project.project_cost_code,
```

Comment:
Is it worth doing `project&.project_cost_code`... not sure if `tube.projects` is able to return nil. If so actually, would need to do `tube.projects&.first` as well 🤔 

Similar question with `tube.studies.first` and `study.uuid`

Reply:
Hmm this is a tricky one.  costCode is a required field and so is the study UUID.  So if they were `nil` there's not a lot I could do to send the message and it might be better for that to bubble up to create an error log, rather than hide it in safe-chaining.

Reply:
OK gotcha. You could throw an explicit exception for it rather than let the nil pointer happen? So you could say something useful like 'expected data is not present - costcode'.

Reply:
Yes fair.  If we leave it as is, the exception will say something about not being able to get attribute project_cost_code on nil object.  If I do the safe-chaining as you suggest and it gets to the point where it's going to encode the message using Avro, you'll get a useful message back from `fastavro` about the field in the message not being allowed to be null.